Nested 'only' parameter

Archelyst opened the pull request:
This allows to specify a nested `only` parameter both at schema instantiation time and during dump() like so:

``` python
class ChildSchema(Schema):
    foo = fields.Field()
    bar = fields.Field()
    baz = fields.Field()

class ParentSchema(Schema):
    bla = fields.Field()
    bli = fields.Field()
    blubb = fields.Nested(ChildSchema)

data = dict(bla=1, bli=2, blubb=dict(foo=42, bar=24, baz=242))

# either when instantiating
sch = ParentSchema(only=('bla', ('blubb', ('foo', 'bar'))))
result = sch.dump(data)

#or when dumping
sch = ParentSchema()
result = sch.dump(data, only=('bla', ('blubb', ('foo', 'bar'))))
```

It is fully backwards compatible. Fixes #402

Archelyst commented:
Concerning the syntax: `only=('bla', ('blubb', ('foo', 'bar')))` means "only bla and blubb and of blubb only foo and bar". If anybody has a better idea, it's probably not hard to change it.

sloria commented:
Thanks for the PR! This is a nice feature. I will review this soon.

Archelyst commented:
Any thoughts about this one?

sloria commented on marshmallow/schema.py:
```diff
@@ -452,7 +452,7 @@ def accessor(cls, func):
     ##### Serialization/Deserialization API #####
 
-    def dump(self, obj, many=None, update_fields=True, **kwargs):
+    def dump(self, obj, many=None, update_fields=True, only=None, **kwargs):
```
I'm not sure about adding `only` as a param here. I think it'd be good to limit the number of params to `dump` and `load`. Otherwise, users might expect all `Meta` options (e.g. `only`, `exclude`, `include`) to be parameters.

Archelyst replied:
Ok, that's a good point. I'll see, if I can implement my use cases without this. What's the rationale for specifying `many` here? There are three and a half places where the schema can be configured (Meta, **init**, and dump/load); I don't see the logic yet, what goes where and why.

sloria commented:
How well does this approach generalize to `exclude`? This feature was requested in https://github.com/marshmallow-code/marshmallow/issues/151 .

sloria commented on marshmallow/schema.py:
```diff
@@ -668,6 +686,11 @@ def _update_fields(self, obj=None, many=False):
+        # Set nested only parameters overriding the only parameters declared on the fields
+        for field_name, field in [(fn, f) for fn, f in ret.items() if hasattr(f, 'only')]:
+            field.only_explicit = None
+            if only_nested and field_name in only_nested:
```
Minor: I think this can be simplified to `if field_name in only_nested:` , since `only_nested` will always be a `dict`.

sloria commented:
Initial pass finished. :+1: Sorry for the delay on my review; things have been busy at work.

sloria commented:
Also, I am bit concerned about the complexity this feature adds to the codebase, which makes me hesitant to merge this. I welcome feedback on whether this is a necessary feature.

Archelyst commented:
> How well does this approach generalize to exclude? This feature was requested in #151 .

I think we could implement the same mechanism for `exclude`, however, from what I read in #151, a nested `only` parameter would suffice. It's just the inverse, it would probably be not too hard to convert any exclude to an only.

> Initial pass finished. :+1: Sorry for the delay on my review; things have been busy at work.

Thanks for the review, man. You've put so much work into this library from which I benefit greatly. No need to apologize :)

Archelyst commented:
> Also, I am bit concerned about the complexity this feature adds to the codebase, which makes me hesitant to merge this. I welcome feedback on whether this is a necessary feature.

Hm, you're right, the whole part the with figuring out which fields to render is a little tricky. However, the addition at hand is mostly encapsulated in one function. I think, "nested only" is worth the extra complexity.

You've probably seen my PR for marshmallow-jsonapi. IMO it would be big and increase adoption a lot if marshmallow would fully implement jsonapi (as much as serialization is concerned). One part of that is choosing the (nested) fields dynamically. You don't happen to come to Europe any time soon? Would be awesome to chat about how we could approach this. Maybe a little hackaton? :)

dscerri commented:
Thanks for all your hard work, this library has been of great benefit to me!

I agree that this adds some complexity to the codebase, however I think this feature would be very useful. My current solution involves having two parent schemas which only differ in the included fields in the nested schema, and this adds unnecessary complexity. 

An alternative dot notation was proposed in a previous PR (#180), which was maybe a little cleaner.

Archelyst commented:
It's probably not too hard to create a little helper that transforms any (dot) syntax into the easier to process list syntax. However, I don't really like the Syntax in PR #180, because adding several nested fields of one nested item is very verbose: `'first_name, last_name, address.street, address.zip_code, address.city'`. Could rather be `'first_name, last_name, address.fields(street, zip_code, city)'`, that's how facebook does it. Anyhow, we could add parsers for both.

Archelyst commented:
I removed the 'only' parameter from dump(). Now it doesn't look that scary any more :)

deckar01 commented on marshmallow/fields.py:
```diff
@@ -499,14 +500,14 @@ def __init__(self, cls_or_instance, **kwargs):
-                                           'must be a subclass of '
-                                           'marshmallow.base.FieldABC')
+                                 'must be a subclass of '
+                                 'marshmallow.base.FieldABC')
```
Unnecessary formatting change.

Making formatting changes to code you didn't functionally modify makes a PR much more difficult to review.

I would recommend removing unnecessary formatting changes and squashing most of these commits since this PR has churned so much.

Archelyst replied:
I'll definitely squash the commits once everything is settled here.

deckar01 commented on marshmallow/fields.py:
```diff
@@ -499,14 +500,14 @@ def __init__(self, cls_or_instance, **kwargs):
-                                           'elements must be of type '
-                                           'marshmallow.base.FieldABC')
+                                 'elements must be of type '
+                                 'marshmallow.base.FieldABC')
```
Unnecessary formatting change.

deckar01 commented on marshmallow/marshalling.py:
```diff
@@ -118,9 +118,9 @@ def serialize(self, obj, fields_dict, many=False,
-                    for idx, d in enumerate(obj)]
+                                  dict_class=dict_class, accessor=accessor,
+                                  index=idx, index_errors=index_errors)
+                   for idx, d in enumerate(obj)]
```
Unnecessary formatting change.

deckar01 commented on marshmallow/marshalling.py:
```diff
@@ -233,9 +233,9 @@ def deserialize(self, data, fields_dict, many=False, partial=False,
-                    for idx, d in enumerate(data)]
+                       partial=partial, dict_class=dict_class,
+                       index=idx, index_errors=index_errors)
+                   for idx, d in enumerate(data)]
```
Unnecessary formatting change.

deckar01 commented on marshmallow/schema.py:
```diff
@@ -318,7 +318,7 @@ class Meta:
             of invalid items in a collection.
         - ``load_only``: Tuple or list of fields to exclude from serialized results.
-        - ``dump_only``: Tuple or list of fields to exclude from deserialization
+        - ``dump_only``: Tuple or list of fields to exclude from deserialization.
```
Unnecessary formatting change.

deckar01 commented on marshmallow/schema.py:
```diff
@@ -471,8 +471,8 @@ def dump(self, obj, many=None, update_fields=True, **kwargs):
-                            'many=True to serialize a collection.',
-                            category=DeprecationWarning)
+                          'many=True to serialize a collection.',
+                          category=DeprecationWarning)
```
Unnecessary formatting change.

deckar01 commented on marshmallow/fields.py:
```diff
@@ -938,7 +938,6 @@ def _serialize(self, value, attr, obj):
             return value.isoformat()
         except AttributeError:
             self.fail('format', input=value)
-        return value
```
Why was this removed? It seems unrelated to the feature.

This was returning `value` if an exception other than `AttributeError` occurred.

deckar01 commented on marshmallow/fields.py:
```diff
@@ -915,7 +916,6 @@ def _deserialize(self, value, attr, data):
         """Deserialize an ISO8601-formatted time to a :class:`datetime.time` object."""
         if not value:   # falsy values are invalid
             self.fail('invalid')
-            raise err
```
Why was this removed? It seems unrelated to the feature.

If `err` is not defined, then a separate issue should be created.

Archelyst replied:
Yeah, this could be removed in an act of bureaucracy and properness :) I'm more in favor of the [Boy Scout rule](bureaucracy). But if it helps merging this PR, sure, I can undo all cleanup. (By the way, I was not looking for things to clean up, I just stumbled over them.)

deckar01 replied:
I can't merge this PR, but I would encourage keeping PRs as concise as possible for the sake of the reviewer.

It is my personal preference to keep unrelated changes out of a PR, although it is generally acceptable to include necissary formatting fixes in files you modified in a separate commit.
http://stackoverflow.com/questions/2214480/committing-when-changing-source-formatting

Removing unrelated lines of code without testing the behavior is probably not a good idea. Opening a separate PR would help keep unrelated changes from getting blocked on each other.

sloria replied:
Good suggestions, @deckar01 . 

I definitely prefer to have PRs as concise and cohesive as possible. That said, I'm not a stickler on it, and I appreciate when people clean up the code. People are volunteering their valuable time to make these PRs, so I typically don't force people to do too much git "busywork".

deckar01 replied:
I went ahead and removed comments about formatting nits that didn't have additional discussion.

Thanks for working on this PR @Tim-Erwin.

sloria commented:
I think we should move forward with this PR, as this feature has been requested by a number of users by now.

There are a couple things to address before merging:

1) Syntax: `dotted.notation` or `('nested', ('tuple', 'notation'))`?

@Tim-Erwin makes a fair point about boilerplate here: https://github.com/marshmallow-code/marshmallow/pull/405#issuecomment-191287474 . That said, we already use dotted notation when accessing attributes, so using dotted notation for `only` might be more consistent.

``` python
city = fields.Str(attribute='address.city')
```

I'm still up in the air about this issue.

2) Nested `exclude` parameter

If we allow for this notation with `only`, users will expect the same notation to work with `exclude`, so this should be implemented, too.

deckar01 commented:
Although the dotted syntax is more verbose, it is very literal and therefore very readable. Nested tuple syntax representing a dictionary filter does not seem common and seems like it would be difficult to document clearly.

That said, I understand the desire to express a tree concisely and without duplication, so my proposal would be to bake the dotted syntax into the implementation and docs, then provide a utility method that flattens a nested tuple into its corresponding dotted property chains.

`flatPropertyChain('nested', ('tuple', ('notation')))`

would evaluate to

`('nested', 'tuple.notation')`

> Edit: I updated the example, because I had it wrong, which also made me realize it is a little confusing. It is not obvious that the nesting logic doesn't apply until the second level of tuples. There are also valid tuples that are invalid property chains. `('nested', ('tuple',))` is `('nested.tuple',)` or `('nested', 'tuple')`?

sloria commented:
That's a sound approach, @deckar01. I'm hesitant to support two ways of doing the same thing, though. One compromise would be to support the dotted syntax out-of-the-box, then add an example in the docs and/or examples/ folder using a helper to use the tuple syntax.

deckar01 commented:
One concern that I have is the addition of `Nested.only_explicit`. I'm not completely sure why it is necessary to maintain a separate collection of only fields.

Here is my stab at an implementation: https://github.com/marshmallow-code/marshmallow/compare/dev...deckar01:dotted-nesting-support-402

Is there an issue with applying the `only` and `exclude` options to the nested schema directly?

Archelyst committed:
allow to specify nested 'only' parameter

Archelyst commented:
Ok, I reverted the commit that introduced the formatting and unrelated changes. I'll make a separate PR for that. Also I squashed everything into one commit.

Concerning the syntax, the nested tuple one seems cleaner to me. It's closer to what's actually needed (a tree) and easier to construct programmatically. Also, it seems more concise compared to what I've seen in your tests. I'd go the other way around and provide a utility to transform dotted notation to tuples. However, I wouldn't strongly insist on that. The dotted notation looks fine as well.

I don't get your statement

> It is not obvious that the nesting logic doesn't apply until the second level of tuples.

Could you give an example for that? The programming is straight forward and does not distinguish between first and following levels.

Archelyst commented:
> One concern that I have is the addition of Nested.only_explicit. I'm not completely sure why it is necessary to maintain a separate collection of only fields.

Well, `only` can be used during schema init as well as during `dump()`.

deckar01 commented:
> I don't get your statement [about the ambiguity of nested tuples.]
> Could you give an example for that?

`('a', ('b',), 'c') -> ('a', 'b', 'c')` or invalid?

`('a', ('b', ()), 'c') -> ('a', 'c')` or invalid?

`('a', ('b', 'c')) -> ('a', 'b', 'c')` or invalid?

`('a', ('b', ('c',)), ('b', ('d',))) -> ('a', 'b.c', 'b.d')` but currently seems to overwrite to `('a', 'b.d')`.

Dotted strings don't allow representing the same field in more than one way, and invalid dotted strings error naturally with the existing field name validation logic.

> Well, `only` can be used during schema init as well as during `dump()`.

There is no `only` parameter for `Schema.dump()`.

Archelyst commented:
Ok, I get that there is a way to write invalid tuples.

> > Well, only can be used during schema init as well as during dump().
> 
> There is no only parameter for Schema.dump().

Sorry, that was bs. The reason for `only_explicit` is that you can specify `only` on `NestedField`s which should not be overwritten. Otherwise next time a schema is instantiated, not the `only` param specified on the `NestedField` will be active, but the last one specified by a nested `only` param during schema init.

deckar01 commented:
`update_fields` does seem to suggest `dump` has mutation side effects that can be disabled. It might be a good idea to add a test for this behavior.

Archelyst commented:
Hm, do you think we should accept side effects in `dump`? That seems counterintuitive to me. `update_fields` is IMHO just an implementation detail (performance improvement for `NestedFields`) that's unfortunately visible from the outside.

deckar01 commented:
[The commit that introduced `update_fields`](https://github.com/marshmallow-code/marshmallow/commit/d39790255d00a7520dbf450a67d91e5f34d698c9) seems to prevent the same schema's fields from being updated more than once. Rather than performance, could this be required to keep recursively nested schema from cycling? It was originally undocumented, but later received [documentation](https://github.com/marshmallow-code/marshmallow/commit/90d11fc4813f6604872c0e53933cf3f07c6e39f4).

Something to keep in mind is that the `Nested` field's underlying `__schema` object [is always a schema instance](https://github.com/marshmallow-code/marshmallow/blob/dev/marshmallow/fields.py#L394) not a class, so you shouldn't have to worry about `dump()` mutating any classes, just schema instances that may have been reused to construct `Nested` fields.

Archelyst commented:
`update_fields` is used in the case of `many=True` where it is sufficient to only update the fields once and use that configuration for all items.

`__schema` is an instance, but it shares the fields (instances) with any other instances. So while it is fine to modify the schema instance, we shouldn't modify fields if it's avoidable. And it's avoidable quite easily, why do you object to that?

deckar01 commented:
> we shouldn't modify fields if it's avoidable ... why do you object to that?

I suggest adding a test for this behavior. Technically you are still modifying the fields, but just adding a new property to avoid overriding `only`. 

> `update_fields` does seem to suggest `dump` has mutation side effects that can be disabled. It might be a good idea to add a test for this behavior.

Archelyst commented:
> I suggest adding a test for this behavior.

Which behavior and what should the test test?

> Technically you are still modifying the fields, but just adding a new property to avoid overriding only.

That's true, but at least one shouldn't override settings a user sets with an operation like `dump()`.

My favorite solution would be to distinguish between fields on a schema class and field instances on a schema instance.

deckar01 commented:
I would test the purpose of `only_explicit` by dumping then asserting that fields reused on another schema have not mutated unexpectedly, and the other schema's dump is not polluted by the first.

It seems that if a `only_explicit` property is not overwritten due to a nested `only` in a second dump, the previous value of `only_explicit` will remain on the field and affect the second dump.

It seems like a lot of complexity has risen out of adding a property to a field that is later inspected by the marshaller to filter fields. I don't think it should be the responsibility of a field to "remember" that it should be dumped. Assigning state info to a persistent field property as a result of a schema operation seems like a poor design choice and `only_explicit` is repeating this pattern.

deckar01 commented:
After looking deeper into it, modifying field instances is not an issue across schema instances, because the `Schema` constructor creates a [deep copy of the `declared_fields` class attribute](https://github.com/marshmallow-code/marshmallow/blob/4c2354c/marshmallow/schema.py#L329) for the instance. It still could be an issue for the field state leaking across multiple `dump()` calls with different arguments on the same schema instance.

Archelyst commented:
Thanks for analyzing this so thoroughly. I missed the deep copy, will make a test like you suggested as soon as I have some time. 

> It still could be an issue for the field state leaking across multiple dump() calls with different arguments on the same schema instance.

Do you have a specific scenario in mind? I'd include that in the tests then.

@sloria , do we have a decision on the syntax and whether the same should be possible with `exclude`? I'd really like to finish this soon.

deckar01 commented:
> It still could be an issue for the field state leaking across multiple dump() calls with different arguments on the same schema instance.

Since `dump()` doesn't accept an `only` parameter, and field instances aren't shared between schema instances, there is no new field state that is created between `dump()` calls, so it appears that `only_explicit` is unnecessary, an `only` could be applied to the nested field directly.

It is a little misleading that `_update_fields()` is called when the schema is initialized (without a data object), then again in every `dump()` (with the data object it is serializing). Nearly everything that `_update_fields()` does during initialization is repeated unnecessarily during a `dump()` (mainly `__set_field_attrs()`).

That said I think this feature can safely be implemented without modifying `fields.py`. Since we should think of modifications to field instances in `_update_fields()` as initialization logic (which are never changed during a dump), so modifying `only` on a `Nested` field during `_update_fields()` semantically is the same as initializing it with the nested `only` fields supplied to the parent field.

Archelyst commented:
I agree. The `_update_fields` is a little odd, especially because it's really nothing that should be used form the outside. Well, that's not our concern here (could be addressed for v3) and it seems we can live without `only_explicit`.

deckar01 commented:
Now that I am thinking about it in the context of initialization logic, it would probably be better to strip off the nested field names from `only` and apply them to the nested fields in a private method that is called directly by the `Schema` constructor.  This would avoid introducing additional complexity into the `_update_fields()` method and more explicitly indicate that the `only` parameter is used to override the construction (deep copy) of nested field instances.

Archelyst commented:
@sloria , let's move on with this. @deckar01 has a [PR with the dot notation](https://github.com/marshmallow-code/marshmallow/pull/468), while I don't agree with the approach (modifying `opts`), it's cleanly implemented and has support for `exclude`. If you wanna go that way, feel free to close this PR and then please merge his. I'd be happy to modify my code here to use the dot-notation and support `exclude` as well. It just would be really nice to come to an end here.

sloria commented:
I've gone ahead and merged #468 . Thanks so much for your work on this.
